make/test.mk

```diff
@@ -0,0 +1,347 @@
+#-------------------------------------------------------------------------------
+
+# Prints the total coverage of a given package.
+# The total coverage is printed as the last argument in the
+# output of "go tool cover". If the requested test name (first argument)
+# Is *, then unit, integration  tests will be combined
+define print-package-coverage
+$(eval TEST_NAME:=$(1))
+$(eval PACKAGE_NAME:=$(2))
+$(eval COV_FILE:="$(COV_DIR)/$(PACKAGE_NAME)/coverage.$(TEST_NAME).mode-$(COVERAGE_MODE)")
+ @if [ "$(TEST_NAME)" == "*" ]; then \
+  UNIT_FILE=`echo $(COV_FILE) | sed 's/*/unit/'`; \
+  INTEGRATON_FILE=`echo $(COV_FILE) | sed 's/*/integration/'`; \
+  COV_FILE=`echo $(COV_FILE) | sed 's/*/combined/'`; \
+	if [ ! -e $${UNIT_FILE} ]; then \
+		COV_FILE=$${INTEGRATION_FILE}; \
+	else \
+		if [ ! -e $${INTEGRATION_FILE} ]; then \
+			COV_FILE=$${UNIT_FILE}; \
+		else \
+			$(GOCOVMERGE_BIN) $${UNIT_FILE} $${INTEGRATION_FILE} > $${COV_FILE}; \
+		fi; \
+	fi; \
+else \
+	COV_FILE=$(COV_FILE); \
+fi; \
+if [ -e "$${COV_FILE}" ]; then \
+	VAL=`go tool cover -func=$${COV_FILE} \
+		| grep '^total:' \
+		| grep '\S\+$$' -o \
+		| sed 's/%//'`; \
+	printf "%-80s %#5.2f%%\n" "$(PACKAGE_NAME)" "$${VAL}"; \
+else \
+	printf "%-80s %6s\n" "$(PACKAGE_NAME)" "n/a"; \
+fi
+endef
+
+# Iterates over every package and prints its test coverage
+# for a given test name ("unit").
+define package-coverage
+$(eval TEST_NAME:=$(1))
+@printf "\n\nPackage coverage:\n"
+$(eval TEST_PACKAGES:=$(shell go list ./... | grep -v $(ALL_PKGS_EXCLUDE_PATTERN)))
+$(foreach package, $(TEST_PACKAGES), $(call print-package-coverage,$(TEST_NAME),$(package)))
+endef
+
+$(COV_PATH_OVERALL): $(GOCOVMERGE_BIN)
+	@$(GOCOVMERGE_BIN) $(COV_PATH_UNIT) > $(COV_PATH_OVERALL)
+
+# Console coverage output:
+
+# First parameter: file to do in-place replacement with.
+define cleanup-coverage-file
+@sed -i '/.*\/sqlbindata\.go.*/d' $(1)
+@sed -i '/.*\/confbindata\.go.*/d' $(1)
+endef
+
+# Output coverage profile information for each function (only based on unit-tests).
+# Re-runs unit-tests if coverage information is outdated.
+.PHONY: coverage-unit
+## Runs the unit tests with producing coverage files for each package
+coverage-unit: prebuild-check $(COV_PATH_UNIT)
+	$(call cleanup-coverage-file,$(COV_PATH_UNIT))
+	@go tool cover -func=$(COV_PATH_UNIT)
+	$(call package-coverage,unit)
+
+
+.PHONY: coverage-all
+# Output coverage profile information for each function.
+# Re-runs unit- and integration-tests if coverage information is outdated.
+coverage-all: prebuild-check clean-coverage-overall $(COV_PATH_OVERALL)
+	$(call cleanup-coverage-file,$(COV_PATH_OVERALL))
+	@go tool cover -func=$(COV_PATH_OVERALL)
+	$(call package-coverage,*)
+
+# HTML coverage output:
+
+# Generate HTML representation (and show in browser) of coverage profile (based on unit tests).
+# Re-runs unit tests if coverage information is outdated.
+.PHONY: coverage-unit-html
```

Comment:
do we need the HTML output? (I've never used it, personally.)

Reply:
Removed in https://github.com/redhat-developer/git-service/pull/2/commits/f75c256c954915418349b8e6c2bb818403b54a6e

Reply:
thanks!